### PR TITLE
Adding in Dirichlet allocation partitioner

### PR DESCRIPTION
**Pull request by emersodb**
# PR Type
Feature

# Short Description

Clickup Ticket(s): [Link](https://app.clickup.com/t/8689qhwfz)

This PR adds in label-based Dirichlet partitioning. In the Dirichlet sampler that we have, we take an existing dataset and subsample the datapoints according to a Dirichlet distribution with a given $\beta$. In this partitioning example, we take an existing dataset and **split** it into pieces by iterating through the labels and use a Dirichlet distribution to assign a portion of those labels to each client. These two approaches are slightly different and both have been used in literature. 

However, the later is used in two benchmarks ([1](https://arxiv.org/pdf/2206.03655), [2](https://www.computer.org/csdl/journal/oj/2024/01/10316561/1S2UbvQk5Tq)), which, for future experiments, we'd like to reproduce.

# Tests Added

Added a test to ensure that the partitioning is happening as expected and has the right relationship to the Dirichlet $\beta$ parameter.

### CODE REVIEW

File: fl4health/clients/nnunet_client.py [hunk 1]
```diff
@@ -201,7 +201,11 @@ def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
         # Set the number of processes for each dataloader.
         if self.n_dataload_proc is None:
             # Nnunet default is 12 or max cpu's. We decrease max by 1 just in case
-            self.n_dataload_proc = min(12, len(os.sched_getaffinity(0)) - 1)
+            # NOTE: The type: ignore here is to skip issues where a local operating system is not compatible
+            # with sched_getaffinity (older versions of MacOS, for example). The code still won't run but mypy won't
+            # complain. Workarounds like using os.cpu_count(), while not exactly the same, are possible.
+            self.n_dataload_proc = min(12, len(os.sched_getaffinity(0)) - 1)  # type: ignore
+
```
Comment (sanaAyrml) [inline fl4health/clients/nnunet_client.py:208]: Good call, I was also keep getting error on my laptop for this.

File: fl4health/utils/dataset.py [hunk 2]
```diff
@@ -103,3 +102,55 @@ def __getitem__(self, index: int) -> Tuple[Dict[str, torch.Tensor], torch.Tensor
...
+        return data, target
+
+    def __len__(self) -> int:
+        return len(self.data)
+
+
+D = TypeVar("D", bound=Union[TensorDataset, DictionaryDataset])
+
+
+def select_by_indices(dataset: D, selected_indices: torch.Tensor) -> D:
```
Comment (sanaAyrml) [inline fl4health/utils/dataset.py:136]: Is there a specific reason we are moving select_by_indices out of LabelBasedSampler?

File: fl4health/utils/dataset.py [hunk 3]
```diff
@@ -103,3 +102,55 @@ def __getitem__(self, index: int) -> Tuple[Dict[str, torch.Tensor], torch.Tensor
...
+    if isinstance(dataset, TensorDataset):
+        modified_dataset = copy.deepcopy(dataset)
```
Comment (sanaAyrml) [inline fl4health/utils/dataset.py:138]: I think the reason to do deepcopy here because it would get a new memory than directly changing input dataset, so we can apply selection multiple times and wouldn't modify the original one. But should not we do the same for DictionaryDataset as well in line 146?

File: fl4health/clients/nnunet_client.py (hunk 1 above)
Comment (emersodb) [reply fl4health/clients/nnunet_client.py:208]: Yeah. CPU checking is a bit fluky across operating systems. We'll just punt this for now.

File: fl4health/utils/dataset.py (hunk 2 above)
Comment (sanaAyrml) [reply fl4health/utils/dataset.py:136]: Oh I see you want to also use it in partitioner. 👍

Comment (emersodb) [reply fl4health/utils/dataset.py:136]: Good question. I extracted the function because it doesn't use any class specific variables and it's actually agnostic of the fact that you're using any kind of label based sampler. That is, this can be used by any type of sampler (say random sampler for a trivial example) as long as you generate indicates to select in the dataset.

File: fl4health/utils/dataset.py (hunk 3 above)
Comment (emersodb) [reply fl4health/utils/dataset.py:138]: I'm not 100% sure. So correct me if I'm wrong here. However, in the original implementation of `select_by_indices` we were making modifications to the `dataset.targets` and `dataset.data` in-place, which resulted in changes to the original object. Here, we're creating a new DictionaryDataset with these values on line 154. So the original dataset should still be intact and unchanged.

File: fl4health/utils/partitioners.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import torch
+from flwr.common.logger import log
+
+from fl4health.utils.dataset import DictionaryDataset, TensorDataset, select_by_indices
+
+T = TypeVar("T")
+D = TypeVar("D", bound=Union[TensorDataset, DictionaryDataset])
+
+
+class LabelBasedDirichletAllocation:
```
Comment (sanaAyrml) [inline fl4health/utils/partitioners.py:15]: In order to be consistent with DirichletLabelBasedSampler maybe we can rename this as DirichletLabelBasedAllocation?

Comment (emersodb) [reply fl4health/utils/partitioners.py:15]: Will do!

File: fl4health/utils/dataset.py (hunk 3 above)
Comment (sanaAyrml) [reply fl4health/utils/dataset.py:138]: Oh I see in the last line we do that. Yes yes then it should be fine.

**Review by sanaAyrml — APPROVED** [review]
Everything looks great, and the current Dirichlet allocation implementation aligns well with what I developed for my master's project. One suggestion would be to include an example that demonstrates how to use the partitioning module. While I was able to figure it out by examining the test functions, having a dedicated example would make it easier for users to understand without needing to refer to the tests.

**Comment by emersodb** [comment]
> Everything looks great, and the current Dirichlet allocation implementation aligns well with what I developed for my master's project. One suggestion would be to include an example that demonstrates how to use the partitioning module. While I was able to figure it out by examining the test functions, having a dedicated example would make it easier for users to understand without needing to refer to the tests.

Yup, will do. I won't implement a full example in our examples folder, since this is a fairly small piece, but I will add a code snippet in the class documentation that gives a basic example 🙂
